proto-lens/src/Data/ProtoLens/Encoding/Bytes.hs

```diff
@@ -76,14 +80,43 @@ putBytes = Builder.byteString
 -- VarInts are inherently unsigned; there are different ways of encoding
 -- negative numbers for int32/64 and sint32/64.
 getVarInt :: Parser Word64
-getVarInt = loop 1 0
+getVarInt = loopStart 1 0
   where
-    loop !s !n = do
+    loopStart !s !n = do
         b <- getWord8
-        let n' = n + s * fromIntegral (b .&. 127)
-        if (b .&. 128) == 0
-            then return $! n'
-            else loop (128*s) n'
+        getVarIntLoopFinish n s b loopStart
+
+-- Same as getVarInt but reads from a Handle
+getVarIntH :: Handle -> IO Word64
+getVarIntH h = do
+    buf <- malloc
+    let loopStart !s !n = do
+          _ <- hGetBuf h buf 1
```

Comment:
You should check the result of `hGetBuf`.  It might return a value of zero at EOF, in which case `b` would be undefined.  At worst case, that would mean this could loop forever.

Reply:
Good point! What would be a reasonable way to fail gracefully in case `hGetBuf` returns `0`?

Reply:
Hm, after thinking about it: since the function that ultimately calls this one returns `IO (Either String ...)`, you should probably return a similar type from this function: `IO (Either String Word64)`.

Maybe it would make sense to simplify the internal error-passing via `ExceptT String IO`.  Up to you.